### ISSUE TITLE
Failed to build gem native extension

### ISSUE THREAD

**Issue by midnight-wonderer** [konsolebox/digest-kangarootwelve-ruby#1]
Hello,
Just playing around with this thing.  

I include the gem into a project via bundler, and I choose the avx2 variant by doing:
~~~bash
bundle config build.digest-kangarootwelve --with-target=avx2
bundle install
~~~

This produces compilation errors:  
~~~
Installing digest-kangarootwelve 0.4.0 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /home/root/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/digest-kangarootwelve-0.4.0/ext/digest/kangarootwelve
/home/root/.rbenv/versions/3.0.3/bin/ruby -I /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0 -r ./siteconf20220212-46354-s0y8h3.rb extconf.rb --with-target\=avx2
creating Makefile

current directory: /home/root/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/digest-kangarootwelve-0.4.0/ext/digest/kangarootwelve
make DESTDIR\= clean

current directory: /home/root/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/digest-kangarootwelve-0.4.0/ext/digest/kangarootwelve
make DESTDIR\=
compiling ./targets/avx2/KangarooTwelve.c
compiling ./targets/avx2/KeccakP-1600-times2-SIMD128.c
In file included from ./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:28,
                 from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c: In function ‘KeccakP1600times2_PermuteAll_24rounds’:
/usr/lib/gcc/x86_64-redhat-linux/11/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
In file included from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:59:37: note: called from here
   59 |         #define ROL64in128_56(a)    _mm_shuffle_epi8(a, CONST128(rho56))
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:567:11: note: in expansion of macro ‘ROL64in128_56’
  567 |     Bmu = ROL64in128_56(A##so); \
      |           ^~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600/common/KeccakP-1600-unrolling.macros:237:9: note: in expansion of macro ‘thetaRhoPiChiIotaPrepareTheta’
  237 |         thetaRhoPiChiIotaPrepareTheta(i+1, E, A) \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:811:5: note: in expansion of macro ‘rounds24’
  811 |     rounds24
      |     ^~~~~~~~
In file included from ./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:28,
                 from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
/usr/lib/gcc/x86_64-redhat-linux/11/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
In file included from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:58:37: note: called from here
   58 |         #define ROL64in128_8(a)     _mm_shuffle_epi8(a, CONST128(rho8))
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:542:11: note: in expansion of macro ‘ROL64in128_8’
  542 |     Bko = ROL64in128_8(A##mu); \
      |           ^~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600/common/KeccakP-1600-unrolling.macros:237:9: note: in expansion of macro ‘thetaRhoPiChiIotaPrepareTheta’
  237 |         thetaRhoPiChiIotaPrepareTheta(i+1, E, A) \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:811:5: note: in expansion of macro ‘rounds24’
  811 |     rounds24
      |     ^~~~~~~~
In file included from ./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:28,
                 from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
/usr/lib/gcc/x86_64-redhat-linux/11/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
In file included from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:59:37: note: called from here
   59 |         #define ROL64in128_56(a)    _mm_shuffle_epi8(a, CONST128(rho56))
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:567:11: note: in expansion of macro ‘ROL64in128_56’
  567 |     Bmu = ROL64in128_56(A##so); \
      |           ^~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600/common/KeccakP-1600-unrolling.macros:236:9: note: in expansion of macro ‘thetaRhoPiChiIotaPrepareTheta’
  236 |         thetaRhoPiChiIotaPrepareTheta(i  , A, E) \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:811:5: note: in expansion of macro ‘rounds24’
  811 |     rounds24
      |     ^~~~~~~~
In file included from ./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:28,
                 from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
/usr/lib/gcc/x86_64-redhat-linux/11/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
In file included from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:58:37: note: called from here
   58 |         #define ROL64in128_8(a)     _mm_shuffle_epi8(a, CONST128(rho8))
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:542:11: note: in expansion of macro ‘ROL64in128_8’
  542 |     Bko = ROL64in128_8(A##mu); \
      |           ^~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600/common/KeccakP-1600-unrolling.macros:236:9: note: in expansion of macro ‘thetaRhoPiChiIotaPrepareTheta’
  236 |         thetaRhoPiChiIotaPrepareTheta(i  , A, E) \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:811:5: note: in expansion of macro ‘rounds24’
  811 |     rounds24
      |     ^~~~~~~~
In file included from ./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:28,
                 from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
/usr/lib/gcc/x86_64-redhat-linux/11/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
In file included from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:58:37: note: called from here
   58 |         #define ROL64in128_8(a)     _mm_shuffle_epi8(a, CONST128(rho8))
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:542:11: note: in expansion of macro ‘ROL64in128_8’
  542 |     Bko = ROL64in128_8(A##mu); \
      |           ^~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600/common/KeccakP-1600-unrolling.macros:236:9: note: in expansion of macro ‘thetaRhoPiChiIotaPrepareTheta’
  236 |         thetaRhoPiChiIotaPrepareTheta(i  , A, E) \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:811:5: note: in expansion of macro ‘rounds24’
  811 |     rounds24
      |     ^~~~~~~~
In file included from ./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:28,
                 from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
/usr/lib/gcc/x86_64-redhat-linux/11/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
In file included from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:59:37: note: called from here
   59 |         #define ROL64in128_56(a)    _mm_shuffle_epi8(a, CONST128(rho56))
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:567:11: note: in expansion of macro ‘ROL64in128_56’
  567 |     Bmu = ROL64in128_56(A##so); \
      |           ^~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600/common/KeccakP-1600-unrolling.macros:236:9: note: in expansion of macro ‘thetaRhoPiChiIotaPrepareTheta’
  236 |         thetaRhoPiChiIotaPrepareTheta(i  , A, E) \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:811:5: note: in expansion of macro ‘rounds24’
  811 |     rounds24
      |     ^~~~~~~~
In file included from ./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:28,
                 from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
/usr/lib/gcc/x86_64-redhat-linux/11/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
In file included from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:58:37: note: called from here
   58 |         #define ROL64in128_8(a)     _mm_shuffle_epi8(a, CONST128(rho8))
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:542:11: note: in expansion of macro ‘ROL64in128_8’
  542 |     Bko = ROL64in128_8(A##mu); \
      |           ^~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600/common/KeccakP-1600-unrolling.macros:237:9: note: in expansion of macro ‘thetaRhoPiChiIotaPrepareTheta’
  237 |         thetaRhoPiChiIotaPrepareTheta(i+1, E, A) \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:811:5: note: in expansion of macro ‘rounds24’
  811 |     rounds24
      |     ^~~~~~~~
In file included from ./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:28,
                 from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
/usr/lib/gcc/x86_64-redhat-linux/11/include/tmmintrin.h:136:1: error: inlining failed in call to ‘always_inline’ ‘_mm_shuffle_epi8’: target specific option mismatch
  136 | _mm_shuffle_epi8 (__m128i __X, __m128i __Y)
      | ^~~~~~~~~~~~~~~~
In file included from ./targets/avx2/KeccakP-1600-times2-SIMD128.c:1:
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:59:37: note: called from here
   59 |         #define ROL64in128_56(a)    _mm_shuffle_epi8(a, CONST128(rho56))
      |                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:567:11: note: in expansion of macro ‘ROL64in128_56’
  567 |     Bmu = ROL64in128_56(A##so); \
      |           ^~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600/common/KeccakP-1600-unrolling.macros:237:9: note: in expansion of macro ‘thetaRhoPiChiIotaPrepareTheta’
  237 |         thetaRhoPiChiIotaPrepareTheta(i+1, E, A) \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./targets/avx2/../../XKCP/lib/low/KeccakP-1600-times2/SIMD128/KeccakP-1600-times2-SIMD128.c:811:5: note: in expansion of macro ‘rounds24’
  811 |     rounds24
      |     ^~~~~~~~
./targets/avx2/KeccakP-1600-times2-SIMD128.c: At top level:
cc1: note: unrecognized command-line option ‘-Wno-self-assign’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-parentheses-equality’ may have been intended to silence earlier diagnostics
cc1: note: unrecognized command-line option ‘-Wno-constant-logical-operand’ may have been intended to silence earlier diagnostics
make: *** [Makefile:245: KeccakP-1600-times2-SIMD128.o] Error 1

make failed, exit code 2

Gem files will remain installed in /home/root/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/digest-kangarootwelve-0.4.0 for inspection.
Results logged to /home/root/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/extensions/x86_64-linux/3.0.0/digest-kangarootwelve-0.4.0/gem_make.out

  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/rubygems/ext/builder.rb:93:in `run'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/rubygems/ext/builder.rb:44:in `block in make'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/rubygems/ext/builder.rb:36:in `each'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/rubygems/ext/builder.rb:36:in `make'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/rubygems/ext/ext_conf_builder.rb:63:in `block in build'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/tempfile.rb:317:in `open'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/rubygems/ext/ext_conf_builder.rb:26:in `build'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/rubygems/ext/builder.rb:159:in `build_extension'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/rubygems/ext/builder.rb:193:in `block in build_extensions'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/rubygems/ext/builder.rb:190:in `each'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/rubygems/ext/builder.rb:190:in `build_extensions'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/rubygems/installer.rb:845:in `build_extensions'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/rubygems_gem_installer.rb:71:in `build_extensions'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/rubygems_gem_installer.rb:28:in `install'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/source/rubygems.rb:200:in `install'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/installer/gem_installer.rb:54:in `install'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/installer/gem_installer.rb:59:in `block in install_with_settings'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/rubygems_integration.rb:575:in `install_with_build_args'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/installer/gem_installer.rb:59:in `install_with_settings'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/installer/gem_installer.rb:16:in `install_from_spec'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/installer/parallel_installer.rb:186:in `do_install'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/installer/parallel_installer.rb:177:in `block in worker_pool'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/worker.rb:62:in `apply_func'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/worker.rb:57:in `block in process_queue'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/worker.rb:54:in `loop'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/worker.rb:54:in `process_queue'
  /home/root/.rbenv/versions/3.0.3/lib/ruby/3.0.0/bundler/worker.rb:91:in `block (2 levels) in create_threads'

An error occurred while installing digest-kangarootwelve (0.4.0), and Bundler cannot continue.

In Gemfile:
  digest-kangarootwelve
~~~

Do I need any other flags or environment variables to build the gem?

**Comment by konsolebox** [comment]
Thanks for the report.  I'll look into it.  It's odd that the equal signs are backslash escaped.

**Comment by konsolebox** [comment]
Looks like avx2 target doesn't compile.  I'll need time and focus to know the cause.  Meanwhile you can try downgrading to avx temporarily.  Also make sure your platform supports the target.

**Comment by midnight-wonderer** [comment]
Thank you for your swift reply, but for me, AVX is still the same issue.  
I figured that it is something to do with bundler config.

It seems we need to pass CPU architecture to GCC, which bundler doesn't, by default.

This works:
~~~bash
bundle config --local build.digest-kangarootwelve --with-target=avx2 --with-cflags=\"-march=broadwell\"
bundle install
~~~

I planned to run this on the Broadwell server; I need to pass `-march` manually or the bundler will set the flag for generic compilation.
This works on both AVX and AVX2, you probably want to mention it in the readme.

**Comment by konsolebox** [comment]
That's nice.  Thank you for finding a solution and the cause.  Yes I will mention it in readme.

**Comment by midnight-wonderer** [comment]
I found out that the native extension does not link properly.
Even if I can compile the gem, it will raise a fatal error at runtime.

~~~
irb(main):001:0> ::Digest::KangarooTwelve[32].digest("abc")
symbol lookup error: /home/root/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/digest-kangarootwelve-0.4.0/lib/digest/kangarootwelve.so: undefined symbol: KeccakP1600_Initialize
~~~

This will make MRI exit with errors.

I understand it is working on your end. But, in the end, I decided to release my own binding; your setup is too complex for me.
https://github.com/the-cave/sleeping-kangaroo12
https://rubygems.org/gems/sleeping_kangaroo12

The fundamental choices are greatly different, so I don't think it is proper as a pull request.

Currently, I am the only one testing out my gem. Not sure if it works on your machine.  
If so, feel free to pick what you are interested in to this repository.

**Comment by konsolebox** [comment]
It turns out that I didn't include the `.s` files for compilation, which `avx2` target also has.  `create_makefile` also doesn't include `.s` or `.S` files when generating the sources list so I generated and specified it manually through the `$srcs` variable.

I pushed a new gem which is version 0.4.2.  I also released an earlier version (0.4.1) which I "yanked" because it still didn't work.  I mistested.

I tested 0.4.2 in another machine which apparently has `avx2` support and it works there.

I might look at your code too some time thank you.
